### ISSUE TITLE
urljoining an empty query string doesn't clear query string

### ISSUE THREAD

**Issue by c29b6996-d572-4c17-906e-e392d8cb0cd9** [python/cpython#76960]
BPO | [32779](https://bugs.python.org/issue32779)
--- | :---
Nosy | @orsenthil, @asvetlov, @thetorpedodog, @iritkatriel
PRs | <li>python/cpython#5645</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-02-06.04:48:49.412>
labels = ['type-bug', 'library', '3.9', '3.10', '3.11']
title = "urljoining an empty query string doesn't clear query string"
updated_at = <Date 2021-05-28.14:36:47.327>
user = 'https://github.com/thetorpedodog'
```

bugs.python.org fields:
```python
activity = <Date 2021-05-28.14:36:47.327>
actor = 'pfish'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-02-06.04:48:49.412>
creator = 'pfish'
dependencies = []
files = []
hgrepos = []
issue_num = 32779
keywords = ['patch']
message_count = 7.0
messages = ['311704', '311937', '312201', '312223', '394648', '394649', '394664']
nosy_count = 4.0
nosy_names = ['orsenthil', 'asvetlov', 'pfish', 'iritkatriel']
pr_nums = ['5645']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue32779'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### Linked PRs
* gh-123273

**Comment by orsenthil** [comment]
Yes, this a bug with urljoin in Python. 

I compared the behavior against Ruby, and Golang.

```ruby
require 'uri'

base_url = 'https://www.example.com/?a=b'
relative_url = '?'

url = URI.join(base_url, relative_url).to_s

puts url
```

Output:

> https://www.example.com/?

And go's https://pkg.go.dev/net/url@go1.19beta1#URL.ResolveReference, ResolveReference resolves a URI reference to an absolute URI from an absolute base URI u, per [RFC 3986 Section 5.2](https://rfc-editor.org/rfc/rfc3986.html#section-5.2).

```go
package main

import (
	"fmt"
	"net/url"
)

func main() {
	base, _ := url.Parse("https://example.com/?a=b")
	u, _ := url.Parse("?")
	fmt.Println(base.ResolveReference(u))
}
```

> https://example.com/?

In python

```
Python 3.12.0a7+ (heads/main:4898415df7, Apr 20 2023, 16:46:44) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib.parse
>>> urllib.parse.urljoin("http://www.example.com/?a/b","?")
'http://www.example.com/?a/b'
```

We can treat this as bug and fix it. However, there are very high chances that people are _relying_ on the current no-op behavior, and this change in behavior can cause frameworks to break. This sounds a bit costly to me.

**Comment by serhiy-storchaka** [comment]
This is not only about "?". For example, "?#z" also should set an empty query. There may be also issues with empty authority ("//"), but this is less clear.

Currently `urllib.parse` does not distinguish between undefined (no delimiter) and empty (there is a delimiter) components, both are represented by empty strings (see #99962). To make the behavior RFC3986 compatible we need a new parser. So I refactored the code in #123273, made internal helper functions more RFC3986 compatible. Public functions are now wrappers around them. Later I will add options to make public parsing/unparsing API also optionally RFC3986 compatible.

It seems that there is an error in RFC3986. According to its algorithm, an the relative reference should always override fragment. And since an empty string does not have fragment, it urljoin() with empty string should remove fragment from the base URI. But older RFCs explicitly said that the original base URI should be returned. RFC3986 does not contain such statement explicitly and there is no such case in examples.

As for "http:?" etc, RFC3986 allows to ignore the scheme if it is the same as the scheme in the base URI for compatibility. Current `urllib.parse` parser is not strict according to RFC3986. We may add an option for strict mode, but this is a different issue.

**Comment by orsenthil** [comment]
> Currently urllib.parse does not distinguish between undefined (no delimiter) and empty (there is a delimiter) components, both are represented by empty strings (see https://github.com/python/cpython/issues/99962). To make the behavior RFC3986 compatible we need a new parser. So I refactored the code in https://github.com/python/cpython/pull/123273, made internal helper functions more RFC3986 compatible.

Thanks for explaining this clearly and taking this approach, @serhiy-storchaka . 

My one feedback is, we make this change early in the cycle of release so that we give ample time for many tools that rely on this underlying parser to detect the behavior change (if they relied on a previous one).

**Comment by serhiy-storchaka** [comment]
Yet one bug (and perhaps more serious one) -- removing trailing semicolon from the path.

```pycon
>>> urljoin('http:/a/b/c/d;#f', '#z')
'http:///a/b/c/d#z'
>>> urldefrag('http:/a/b/c/d;#f')
DefragResult(url='http:///a/b/c/d', fragment='f')
```

RFC3986 does not define "params". The URI consists of 5 components: scheme, authority, path, query, fragment. `'/b/c/d;'` is the path, and the `urllib` functions mutilate it.

This is why I think that it is worth to backport the solution.

**Comment by orsenthil** [comment]
> RFC3986 does not define "params".
True. And when I dug a bit into this topic yesterday. Using params is highly discouraged now, and I think, very old Sun Servers and FTPs used a params for storing _session state_ 

The backport is fine. But it carries the risk of breaking some applications. Calling it out and early backport is alright IMO.

**Comment by serhiy-storchaka** [comment]
Should this change be backported?

**Comment by hugovk** [comment]
Triage: if we're not backporting, let's close this.
